src/actions/groups.cr

```diff
@@ -1,34 +1,171 @@
+      specified request body at #{context.request.path}
+      HERE
+  end
+
   extend self
   include RouteHelpers
   relative_post do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Create
-    # TODO: create a group based on the response body
-
+    body = context.request.body
+    raise BadRequest.new context, "One must specify a group to add." if body.nil?
+    group = Config::Group.from_json body
+    DppmRestApi.permissions_config.groups << group
+    DppmRestApi.permissions_config.sync_to_disk
+    build_json context.response do |response|
+      response.field("successfullyAddedGroup") { group.to_json response }
+    end
   end
-  relative_put "/:id/route" do |context|
+  # Regardless of whether the given path exists on the given group, set the
+  # group's access level on that path to the given access level.
+  relative_put "/:id/route/:path/:access_level" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Update
-    # TODO: add a new route to the group
-
+    group_id_s = context.params.url["id"]
+    group_id = group_id_s.to_i? || raise BadRequest.new context,
+      %[Group ID #{group_id_s} was not a valid integer]
+    # find the group if it already exists in the permissions config
+    index = DppmRestApi.permissions_config
+      .groups
+      .index { |grp| grp.id == group_id }
+    raise NoSuchGroup.new context, group_id_s if index.nil?
+    old_group = DppmRestApi.permissions_config.groups[index]
+    path = URI.unescape context.params.url["path"]
+    access_level = Access.parse?(context.params.url["access_level"])
+    raise InvalidAccessLevel.new context if access_level.nil?
+    name = context.params
+      .query["name"]?
+      .try { |param| URI.unescape param } || old_group.name
+    # Create the group if it doesn't already exist
+    group = Config::Group.new id: group_id, name: name,
+      permissions: old_group.permissions
+    # Create a query parameters object (since the default isn't nil, it's an
+    # empty Hash) if group doesn't already have a path with query params on
+    # it.
+    existing_params = group.permissions[path]?
+      .try &.query_parameters || {} of String => Array(String)
+    # if there were query parameters specified in the body, override
+    # existing values
+    existing_params.merge! query_params_in_body?(of: context) do |_path, existing, new_val|
+      existing + new_val
+    end
+    # Override the existing permissions
+    group.permissions[path] = Config::Route.new access_level, existing_params
+    # and overwrite the current value.
+    DppmRestApi.permissions_config.groups[index] = group
+    build_json context.response do |response|
+      response.field "successfullyModifiedGroup" do
+        response.object do
+          response.field("from") { old_group.to_json response }
+          response.field("to") { group.to_json response }
```

Comment:
An API response should better have declarative names.
I suggest `source` and `target`